rust/src/meas/gaussian/mod.rs

```diff
@@ -20,92 +24,154 @@ use self::analytic::get_analytic_gaussian_epsilon;
-            .map(|v| T::sample_gaussian(*v, scale, false))
-            .collect())
+        Function::new_fallible(move |arg: &Self::Carrier| {
+            arg.iter()
+                .map(|v| T::sample_gaussian(*v, scale, false))
+                .collect()
+        })
     }
 }
 
+pub trait GaussianMeasure<MI: GaussianMetric>: Measure + Default {
+    type Atom;
+    fn new_forward_map(scale: Self::Atom) -> PrivacyMap<MI, Self>;
+}
 
-pub fn make_base_gaussian<D>(
-    scale: D::Atom
-) -> Fallible<Measurement<D, D, D::Metric, SmoothedMaxDivergence<D::Atom>>>
-    where D: GaussianDomain,
-          D::Atom: 'static + Clone + SampleGaussian + Float + InfCast<f64> + InfSub + InfDiv + CheckNull + InfMul + InfAdd + InfLn + InfSqrt + InfExp + Zero {
+pub trait GaussianMetric: Metric {}
+impl<Q: CheckNull> GaussianMetric for L2Distance<Q> {}
+impl<Q: CheckNull> GaussianMetric for AbsoluteDistance<Q> {}
+
+impl<MI: GaussianMetric, Q> GaussianMeasure<MI> for SmoothedMaxDivergence<Q>
+where
+    MI: Metric<Distance = Q>,
+    Q: 'static
+        + Clone
+        + SampleGaussian
+        + Float
+        + InfCast<f64>
+        + InfSub
+        + InfDiv
+        + CheckNull
+        + InfMul
+        + InfAdd
+        + InfLn
+        + InfSqrt
+        + InfExp
+        + Zero,
+{
+    type Atom = Q;
+    fn new_forward_map(scale: Q) -> PrivacyMap<MI, Self> {
+        PrivacyMap::new_fallible(move |&d_in: &Q| {
+            if d_in.is_sign_negative() {
+                return fallible!(InvalidDistance, "sensitivity must be non-negative");
+            }
+
+            let _2 = Q::inf_cast(2.)?;
+            let additive_gauss_const = Q::inf_cast(ADDITIVE_GAUSS_CONST)?;
+
+            Ok(SMDCurve::new(move |del: &Q| {
+                if !del.is_sign_positive() {
+                    return fallible!(FailedRelation, "delta must be positive");
+                }
+
+                if scale.is_zero() {
+                    return Ok(Q::infinity());
+                }
+
+                let eps = d_in
+                    .inf_mul(
+                        &additive_gauss_const
+                            .inf_add(&_2.inf_mul(&del.recip().inf_ln()?)?)?
+                            .inf_sqrt()?,
+                    )?
+                    .inf_div(&scale)?;
+
+                if eps > Q::one() {
+                    return fallible!(RelationDebug, "The gaussian mechanism has an epsilon of at most one. Epsilon is greater than one at the given delta.");
+                }
+                Ok(eps)
+            }))
+        })
+    }
+}
+
+impl<MI, Q> GaussianMeasure<MI> for ZeroConcentratedDivergence<Q>
+where
+    MI: GaussianMetric<Distance = Q>,
+    Q: 'static + Clone + ExactIntCast<usize> + InfPow + InfDiv,
+{
+    type Atom = Q;
+
+    fn new_forward_map(scale: Q) -> PrivacyMap<MI, Self> {
+        PrivacyMap::new_fallible(move |d_in: &Q| {
+            let _2 = Q::exact_int_cast(2)?;
+            d_in.inf_div(&scale)?.inf_pow(&_2)?.inf_div(&_2)
```

Comment:
Just confirming this is correct: It's (d_in / scale)^2 / 2, not d_in * scale^2 / 2

Reply:
It is correct, but I'll throw in some redundant parens to make it more clear.

`d_in.inf_div(&scale)?` is fully evaluated to type Q before the `inf_pow` trait method from Q is invoked.